### PR TITLE
Use core::time::Duration with TimeDelta as output type only

### PR DESCRIPTION
**Pull request by esheppa**
Setting this as draft while I figure out a good way to split up the commit into manageable units

### CODE REVIEW

**Review by djc — COMMENTED** [review]
Some thoughts/suggestions from an initial round of skimming.

File: src/naive/datetime/mod.rs [hunk 1]
```diff
@@ -737,23 +739,43 @@ impl NaiveDateTime {
...
+    /// # use core::time::Duration;
+    /// let leap = NaiveDate::from_ymd(2015, 6, 30).and_hms_milli(23, 59, 59, 1_500);
+    /// assert_eq!(leap.signed_duration_since(NaiveDate::from_ymd(2015, 6, 30).and_hms(23, 0, 0)),
+    ///            TimeDelta::Forwards(Duration::from_secs(3600) + Duration::from_millis(500)));
+    /// assert_eq!(NaiveDate::from_ymd(2015, 7, 1).and_hms(1, 0, 0).signed_duration_since(leap),
+    ///            TimeDelta::Forwards(Duration::from_secs(3600) - Duration::from_millis(500)));
     /// ```
     pub fn signed_duration_since(self, rhs: NaiveDateTime) -> TimeDelta {
-        self.date.signed_duration_since(rhs.date) + self.time.signed_duration_since(rhs.time)
+        match (self.date.days_since(rhs.date), self.time.signed_duration_since(rhs.time)) {
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/datetime/mod.rs:758]: I've made a new implementation for this with some extra methods on `TimeDelta`

File: src/time_delta.rs [hunk 2]
```diff
@@ -38,42 +35,91 @@ const SECS_PER_DAY: i64 = 86400;
...
 /// ISO 8601 time duration with nanosecond precision.
 ///
 /// This also allows for the negative duration; see individual methods for details.
-#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+#[derive(Clone, Copy, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "rkyv", derive(Archive, Deserialize, Serialize))]
-pub struct TimeDelta {
-    secs: i64,
-    nanos: i32, // Always 0 <= nanos < NANOS_PER_SEC
+pub enum TimeDelta {
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/time_delta.rs:34]: This is now part of bc9709a which just adds the new `TimeDelta` type

File: src/naive/date.rs [hunk 3]
```diff
@@ -1136,22 +1203,23 @@ impl NaiveDate {
...
-    /// assert_eq!(d.checked_sub_signed(TimeDelta::days(-40)),
-    ///            Some(NaiveDate::from_ymd(2015, 10, 15)));
-    /// assert_eq!(d.checked_sub_signed(TimeDelta::days(1_000_000_000)), None);
-    /// assert_eq!(d.checked_sub_signed(TimeDelta::days(-1_000_000_000)), None);
-    /// assert_eq!(NaiveDate::MIN.checked_sub_signed(TimeDelta::days(1)), None);
+    /// assert_eq!(d.checked_sub(Duration::from_secs(24 * 60 * 60 * 1_000_000_000)), None);
+    /// assert_eq!(NaiveDate::MIN.checked_sub(Duration::from_secs(24 * 60 * 60 * 1)), None);
     /// ```
-    pub fn checked_sub_signed(self, rhs: TimeDelta) -> Option<NaiveDate> {
+    #[deprecated(since = "0.5.0", note = "Use Days::new(n)")]
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/date.rs]: Good call, I've now removed these all in 31d352b, left them in prior to that so that the tests pass on all the commits

File: src/naive/date.rs [hunk 4]
```diff
@@ -733,7 +773,20 @@ impl NaiveDate {
             return Some(self);
         }
 
-        i64::try_from(days.0).ok().and_then(|d| self.diff_days(d))
+        let year = self.year();
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/date.rs]: This code is moved and altered from the previous `checked_add_signed`, which is later removed

File: src/naive/date.rs [hunk 5]
```diff
@@ -752,11 +805,24 @@ impl NaiveDate {
             return Some(self);
         }
 
-        i64::try_from(days.0).ok().and_then(|d| self.diff_days(-d))
-    }
+        let year = self.year();
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/date.rs]: Good call, I've not done this yet but will look into some de duplication here. This was previously handled by just taking the `ops::Neg` of the duration and calling `checked_add_signed` but now we can't do that as the `Duration` is always positive. There are some differences from the `checked_add` impl but it should be possible to consolidate, potentially with helper methods in `internals.rs`

**Comment by esheppa** [comment]
will move f6dceba into the proper commits once I've got the CI passing

File: src/naive/time/mod.rs [hunk 6]
```diff
@@ -657,6 +815,134 @@ impl NaiveTime {
...
+        //          |====>|   |                             |======>|
+        //      `rhs.frac`|========================================>|
+        //          |     |   |        `self - rhs`         |       |
+
+        // importantly, here if the `secs` are larger, then that `NaiveTime` is larger,
+        // no matter what the value of the frac is.
+        // this has the effect that if the smaller time is in a leap second, we can simply remove `SECOND_AS_NANOS`
+        // from the frac, and then calculate as normal
+
+        match self.secs.cmp(&rhs.secs) {
```
Comment (djc) [inline src/naive/time/mod.rs:693]: Why is this related to the `TimeDelta` work?

File: src/month.rs [hunk 7]
```diff
@@ -186,6 +186,66 @@ impl Months {
     }
 }
 
+/// A difference in a number of months, either forwards or backwards.
```
Comment (djc) [inline src/month.rs:189]: This doesn't seem like it necessarily belongs in this PR?

**Review by djc — COMMENTED** [review]
I think the constituent pieces here are looking good. In terms of the commit history, I think instead of leaving the removal of all code related to the `OldTimeDelta` until the end, each commit that adds implementations for `Duration` should also remove the similar implementations for `OldTimeDelta`. That way it's much easier to review if we're adding back equivalent operations.

You also added `DaysDelta` and `MonthsDelta`. It seems like these are additive (that is, they don't exactly replace equivalent functionaly that would be removed with `OldTimeDelta`)? If they are not additive, it would be good to demonstrate that using the same approach (removing code in the same commit as these are added).

File: src/naive/date.rs [hunk 8]
```diff
@@ -120,16 +121,99 @@ impl NaiveWeek {
...
         Self(num)
     }
 }
 
+/// A difference in a number of days, either forwards or backwards.
+///
+/// This type is often returned from fuctions but is generally not used as a parameter.
+/// Instead the inner `Days` must first be extracted and then used. There are helper methods
+/// which can assist with this.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord)]
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/date.rs]: I've now moved this to https://github.com/chronotope/chrono/pull/859

File: src/month.rs (hunk 7 above)
Comment (esheppa) [reply src/month.rs]: I've moved this to https://github.com/chronotope/chrono/pull/859

File: src/naive/time/mod.rs (hunk 6 above)
Comment (esheppa) [reply src/naive/time/mod.rs:696]: This used to rely on the fact that the duration could potentially be negative or positive, now that it can't and so the situation is a bit more  complex and the signed integers are harder to work with.

An option that works which is closer to the original implementation is:

```rust
let mut secs = i64::from(self.secs) - i64::from(rhs.secs);
let mut frac = i64::from(self.frac) - i64::from(rhs.frac);

// `secs` may contain a leap second yet to be counted
let adjust = match self.secs.cmp(&rhs.secs) {
    Ordering::Greater => {
        if rhs.frac >= 1_000_000_000 {
            1
        } else {
            0
        }
    }
    Ordering::Equal => 0,
    Ordering::Less => {
        if self.frac >= 1_000_000_000 {
            -1
        } else {
            0
        }
    }
};

secs += adjust;

while frac < 0 {
    frac += 1_000_000_000;
    secs -= 1;
}
match secs.cmp(&0) {
    Ordering::Equal => {
        TimeDelta::Forwards(Duration::new(0, u32::try_from(frac).unwrap()))
    }
    Ordering::Greater => {
        TimeDelta::Forwards(Duration::new(u64::try_from(secs).unwrap(), u32::try_from(frac).unwrap()))
    }
    Ordering::Less => {
        TimeDelta::Backwards(Duration::new(u64::try_from(secs.abs()).unwrap(), u32::try_from(frac.abs()).unwrap()))
    }
}
```

let me know which you think is better? One benefit of the other way is that it doesn't have any `unwrap`s

File: src/naive/time/mod.rs [hunk 9]
```diff
@@ -565,6 +570,161 @@ impl NaiveTime {
...
+    /// use core::time::Duration;
+    ///
+    /// let from_hms = NaiveTime::from_hms;
+    ///
+    /// assert_eq!(from_hms(3, 4, 5).checked_overflowing_add(Duration::from_secs(60 * 60 * 11)),
+    ///            Some((from_hms(14, 4, 5), Days::new(0))));
+    /// assert_eq!(from_hms(3, 4, 5).checked_overflowing_add(Duration::from_secs(60 * 60 * 23)),
+    ///            Some((from_hms(2, 4, 5), Days::new(1))));
+    /// ```
+    pub fn checked_overflowing_add(&self, rhs: Duration) -> Option<(NaiveTime, Days)> {
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/time/mod.rs]: I've renamed to `overflowing_add_opt` but perhaps there is a better option.

The original intention was to use `checked` because of the `Option` and `overflowing` represents the fact that it is ok for the `Duration` to be longer than the time to the end of the day.

File: src/time_delta.rs [hunk 10]
```diff
@@ -9,16 +9,98 @@
...
+
 #[cfg(any(feature = "std", test))]
 use std::error::Error;
 
 #[cfg(feature = "rkyv")]
 use rkyv::{Archive, Deserialize, Serialize};
 
+#[derive(Clone, Copy, PartialOrd, Ord, Debug)]
+#[cfg_attr(feature = "rkyv", derive(Archive, Deserialize, Serialize))]
+/// A difference (delta) in time build using [`core::time::Duration`] and a direction
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/time_delta.rs:20]: should be resolved in the latest force push

File: src/naive/time/mod.rs [hunk 11]
```diff
@@ -565,6 +570,161 @@ impl NaiveTime {
         (time, -rhs) // safe to negate, rhs is within +/- (2^63 / 1000)
     }
 
+    /// Adds given `TimeDelta` to the current time,
```
[the remark this answers is not in the document]
Comment (esheppa) [reply src/naive/time/mod.rs]: should be resolved in the latest force push

**Comment by djc** [comment]
Don't have time to review this in detail today. One question on my mind: is there some way we can start exposing parts of this change on 0.4.x already? Presumably we could just add the `TimeDelta` type and add methods that return it and/or `Add`/`Sub` impls that take `std::time::Duration`, right?

**Comment by esheppa** [comment]
> can start exposing parts of this change on 0.4.x

This should be quite possible. Almost all the methods have new names as well. The PR is actually more problematic in `main` because we have already used the `TimeDelta` name. The new code without the removals could be quite cleanly applied in `0.4.x` I think.

As an additional thought, what do you think about deprecating the `Add` / `Sub` impls - these don't expose the fallibility anyway (although arguably the fallibility there is primarily overflow which is already tolerated in Add/Sub impls), and then this would make the methods the primary API, meaning we can expose essentially all of this in `0.4.x`

**Comment by djc** [comment]
I think it would be great to first do a PR to 0.4.x which adds new API and deprecates old stuff (I forget if/how well deprecations on `impl` blocks work?) merge that to main and then clean up the old stuff.

**Comment by pitdicker** [comment]
Personally I don't think an enum wrapping an `u64` to pretend to be an signed integer is a good direction. Especially if we are not even using most of the range of that `u64`. `std::time::Duration` is a little more than an `u64`, but the principle is close.

I'm closing this PR for now because it hasn't seen any activity in the last year while it is going to need a lot of work to become merge-able.
